src/persist/src/operators/stream.rs

```diff
@@ -739,114 +742,27 @@ where
-                        // and await_frontier().
-                        pending_futures.push_back((
-                            cap.delayed(cap.time()),
-                            cap.retain_for_output(error_output_port),
-                            write_future,
-                        ));
-                    }
-                });
-
-                // Swing through all pending futures and see if they're ready. Ready futures will
-                // invoke the Activator, which will make sure that we arrive here, even when there
-                // are no changes in the input frontier or new input.
-                let waker = futures_util::task::waker_ref(&activator);
-                let mut context = Context::from_waker(&waker);
-
-                while let Some((cap, error_cap, pending_future)) = pending_futures.front_mut() {
-                    match Pin::new(pending_future).poll(&mut context) {
-                        std::task::Poll::Ready(result) => {
-                            match result {
-                                Ok(seq_no) => {
-                                    log::trace!(
-                                        "In {}, finished writing for time: {}, seq_no: {:?}",
-                                        &operator_name,
-                                        cap.time(),
-                                        seq_no,
-                                    );
-                                }
-                                Err(e) => {
-                                    let mut session = error_output.session(&error_cap);
-                                    let error = format!(
-                                        "In {}, error writing data for time {}: {}",
-                                        &operator_name,
-                                        error_cap.time(),
-                                        e
-                                    );
-                                    log::error!("{}", error);
+        let (pass_through, to_retract) =
+            self.branch(move |_, (_, t, _)| if t < &upper_ts { false } else { true });
+
+        let (retract_oks, errs) = to_retract
+            .map(|(data, time, diff)| (data, time, -diff))
+            .persist(name, write);
+
+        // Introduce a data-dependency between pass_through, and retract_oks even
+        // though in reality they operate on disjoint subsets of data and no true
+        // data dependency exists. This way, we ensure that the frontier doesn't
+        // advance for downstreams users until after all retractions have been
+        // persisted.
+        //
+        // TODO: we could have done this with fewer operators by concatenating
+        // pass_through, to_retract, and retract_oks, as to_retract and retract_oks
+        // cancel out. This approach seemed more isolated and safer.
+        let retract_oks = retract_oks.flat_map(|_| None);
```

Comment:
I guess the one downside of this is that the `flat_map` might still swing through all the retractions?

Reply:
right. we could change `persist` to take a boolean argument that determines whether it passes through the input or just discards it. just thinking out loud through a mitigation

Reply:
added a todo with this note.